src/raftstore/coprocessor/split_check/keys.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use raftstore::store::{util, Msg};
+use rocksdb::DB;
+use util::transport::{RetryableSendCh, Sender};
+
+use super::super::metrics::*;
+use super::super::{Coprocessor, KeyEntry, ObserverContext, SplitCheckObserver, SplitChecker};
+use super::Host;
+
+pub struct Checker {
+    max_keys: u64,
+    split_keys: u64,
+    current_keys: u64,
+    split_key: Option<Vec<u8>>,
+}
+
+impl Checker {
+    pub fn new(max_keys: u64, split_keys: u64) -> Checker {
+        Checker {
+            max_keys,
+            split_keys,
+            current_keys: 0,
+            split_key: None,
+        }
+    }
+}
+
+impl SplitChecker for Checker {
+    fn on_kv(&mut self, _: &mut ObserverContext, row: &KeyEntry) -> bool {
+        if row.is_commit_version() {
+            self.current_keys += 1;
+        }
+        if self.current_keys > self.split_keys && self.split_key.is_none() {
+            self.split_key = Some(row.key().to_vec());
+        }
+        self.current_keys > self.max_keys
+    }
+
+    fn split_key(&mut self) -> Option<Vec<u8>> {
+        if self.current_keys >= self.max_keys {
+            self.split_key.take()
+        } else {
+            None
+        }
+    }
+}
+
+pub struct KeysCheckObserver<C> {
+    region_max_keys: u64,
+    split_keys: u64,
+    ch: RetryableSendCh<Msg, C>,
+}
+
+impl<C: Sender<Msg>> KeysCheckObserver<C> {
+    pub fn new(
+        region_max_keys: u64,
+        split_keys: u64,
+        ch: RetryableSendCh<Msg, C>,
+    ) -> KeysCheckObserver<C> {
+        KeysCheckObserver {
+            region_max_keys,
+            split_keys,
+            ch,
+        }
+    }
+}
+
+impl<C> Coprocessor for KeysCheckObserver<C> {}
+
+impl<C: Sender<Msg> + Send> SplitCheckObserver for KeysCheckObserver<C> {
+    fn add_checker(&self, ctx: &mut ObserverContext, host: &mut Host, engine: &DB) {
+        let region = ctx.region();
+        let region_id = region.get_id();
+        let region_keys = match util::get_region_approximate_keys(engine, region) {
```

Comment:
Not like `get_region_approximate_size`, `get_region_approximate_keys` 's accuracy is very low, this can cause many scan for this checker. You'd better test it.

Reply:
Ok,I'd like add `KeyProperties` to make the accurancy higher

Reply:
Ok, I will do it in another PR.